### ISSUE TITLE
Targeted API improvements by use.

### ISSUE THREAD

**Issue by jni** [scikit-image/scikit-image#4059]
## Description

You might have seen a lightning talk at SciPy 2019 by @costrouc of Quansight about measuring API usage. His work is [here](https://github.com/Quansight-Labs/python-api-inspect) and a blog post discussing it is [here](https://labs.quansight.org/blog/2019/05/python-package-function-usage/).

The data from that work is presented as a public, read-only SQL database. Here's a [query](https://python-api-inspect.aves.io/inspect?sql=++SELECT+key+AS+function%2C+sum%28json_extract%28value%2C+%27%24.count%27%29%29+as+count%0D%0A++FROM+FunctionStats%2C+json_each%28FunctionStats.stats%29%0D%0A++JOIN+File+ON+FunctionStats.id+%3D+File.id%0D%0A++WHERE+FunctionStats.namespace+%3D+%3Anamespace%0D%0A++++AND+File.filename+NOT+LIKE+%27%25%2Fsite-packages%2F%25%27%0D%0A++++AND+File.filename+LIKE+%27%25.ipynb%27%0D%0A++GROUP+BY+key%0D%0A++ORDER+BY+sum%28json_extract%28value%2C+%27%24.count%27%29%29+desc%0D%0A++LIMIT+100&namespace=skimage) of the 100 most-used skimage functions in Jupyter notebooks hosted on GitHub (in a dataset of ~10K repos; I don't know how these were chosen... @costrouc?). I have no SQL skills but we can modify it to suit our purposes (for example, looking in .py files instead of .ipynb files).

If we can, for example, speed up functions in the top-20, we can have a lot more impact than speeding up things near the bottom of that pile.

Another interesting idea, but more challenging, would be to check common parameter settings for each function, as they might inform better defaults in the future.

**Comment by costrouc** [comment]
> In a dataset of ~10K repos; I don't know how these were chosen... @costrouc?

Around [26k repositories](https://python-api-inspect.aves.io/inspect?sql=++SELECT+count%28DISTINCT+project%29+FROM+File) were looked at (included many that are forks). The dataset was taken by using [libraries.io](https://libraries.io/pypi/scikit-image) to get the top 4k repositories each for numpy, scipy, scikit-learn, scikit-image, pytorch, tensorflow, requests, and 14 others. So the dataset is quite biased towards scientific modules. Since I included scikit-image as one of the projects I think that the usage statistics of the api should be accurate. That said I definitely was to expand the repositories that I parse to include other domains in python such as devops and web development.

**Comment by jni** [comment]
@costrouc thanks! Scientific bias is A-OK by us. =P My main question is about what "top 4k" means and whether we could expand these methods to use e.g. GitHub's "used by". We actually care quite a bit about the long tail of users that might be one-off scientific analyses. Their usage might look quite different from that of libraries.

**Comment by sciunto** [comment]
If I got correctly the source used for the data analysis, my concern about bias would rather be that most of the projects are libraries, and not end-user codes, which is likely to be different.

**Comment by costrouc** [comment]
Yeah I agree. What metric do you think could be used to distinguish between end-user codes and libraries?

**Comment by sciunto** [comment]
Nevertheless, being aware of this, it has its interest. I'm not saying that this bias is a killer. ;)

I have no idea  for a metric right away. To get more end-user codes, I'm wondering to which extent it could be possible to analyze source codes provided with scientific publications (can they be collected easily?). That could be a way to get the other side of the coin.

**Comment by jni** [comment]
I think collecting github repos containing notebooks and checking those notebooks is as good as it gets...

**Comment by emmanuelle** [comment]
These statistics are very interesting, thank you @costrouc For the acceleration of functions, we also need to focus on the ones which take a lot of time (say on a 10000 x 10000 image).
